### PR TITLE
Issue 114: add safe iterator for generate pt nowcast matrix list

### LINKED ISSUE #114: Consider using purrr safely or similar try catch in iterators
Related to #106 it might be sensible to handle errors in iterators either safely or by linking errors to the iterator step to help with debugging. 

The downside to using `purrr` is that it adds another dependency but the upside is it can be implemented more easily than a try catch trigger. Generally `purrr` is quite a lightweight dep and is used elsewhere in the `epinowcast` ecosystem. We could also consider using it for other iterators to streamline our code. 

Try catch is nice as it is light weight and it only takes a few lines of code to replicate safely behaviour or similar. However, it adds technical debt to do so. See `scoringutils` for a similar discussion.

### PR DESCRIPTION
**Pull request by kaitejohnson**
## Description

This PR closes #114. It does the following:
- adds a `.safelydoesit` internal function and tests it 
- in `generate_pt_nowcast_mat_list()` uses the suggested wrapper around `tryCatch` to create a safe function, which will return NULL in the list for any element that errors
- adds an informative message in `generate_pt_nowcast_mat_list()` which tells the user how many retrospective nowcasts are NULL out of the number passed in
- adds a step in `estimate_dispersion()` that checks for NULL entries in the `pt_nowcast_mat_list` and removes the corresponding entry from the `trunc_rep_mat_list`. 

## Checklist

- [X] My PR is based on a package issue and I have explicitly linked it.
- [X] I have included the target issue or issues in the PR title in the for Issue(s) *issue-numbers*: PR title
- [X] I have read the [contribution guidelines](https://github.com/epinowcast/.github/blob/main/CONTRIBUTING.md).
- [X] I have tested my changes locally.
- [X] I have added or updated unit tests where necessary.
- [X] I have updated the documentation if required.
- [X] My code follows the established coding standards.
- [X] I have added a news item linked to this PR.
- [X] I have reviewed CI checks for this PR and addressed them as far as I am able.

## Summary by CodeRabbit

- **New Features**
  - Improved error handling during nowcast matrix generation, allowing processing to continue even if some elements fail.
  - Added a safe execution utility to prevent errors from stopping batch processes.

- **Bug Fixes**
  - Enhanced handling of NULL values in dispersion estimation, ensuring robust processing and clear error messages when no valid data is present.

- **Tests**
  - Added tests for safe execution, error handling in nowcast matrix generation, and edge cases involving NULL or zero-valued inputs.

- **Documentation**
  - Added documentation for the new safe execution utility describing its behavior and usage.

### CODE REVIEW

**Comment by seabbs** [comment]
> Provide an informative warning which tells the user how many triangles were indicated to use for estimating dispersion and how many were actually used.

THis one I think (or a message?)

**Comment by seabbs** [comment]
> adds an informative message in generate_pt_nowcast_mat_list() which tells the user how many retrospective nowcasts are NULL out of the number passed in

Does it tell them which ones?

File: R/generate_pt_nowcast_mat_list.R [hunk 1]
```diff
@@ -64,17 +64,48 @@ generate_pt_nowcast_mat_list <- function(reporting_triangle_list,
...
-    generate_pt_nowcast_mat,
+    function(triangle, pmf) {
+      result <- safe_generate_pt_nowcast_mat(
+        reporting_triangle = triangle,
+        delay_pmf = pmf,
+        n = n,
+        max_delay = max_delay
+      )
+      if (!is.null(result$error)) {
+        # Return NULL if there was an error
```
Comment (seabbs) [inline R/generate_pt_nowcast_mat_list.R:81]: here I suggest returning the errors as a warning or message along with the index lof the list. To me this suggests reworking this iterator to work other an index vs the current approach.

**Review by seabbs — CHANGES_REQUESTED** [review]
Some code duplication and suggests for smoothing the user debug path.

File: R/generate_pt_nowcast_mat_list.R (hunk 1 above)
Comment (kaitejohnson) [reply R/generate_pt_nowcast_mat_list.R:81]: Are you suggesting that you want the result of this function to be a nested list, where each indexed element contains a `result` (either NULL or a matrix) and an `error` (either NULL or the nested list of error information)?

Comment (seabbs) [reply R/generate_pt_nowcast_mat_list.R:81]: no I am suggesting printing them as a warning or message. take a look at the scoringutils iterator

Comment (kaitejohnson) [reply R/generate_pt_nowcast_mat_list.R]: I have updated this to provide a message with the indices and the error message corresponding to the call to `generate_pt_nowcast()`

File: R/generate_pt_nowcast_mat_list.R [hunk 2]
```diff
@@ -64,17 +64,48 @@ generate_pt_nowcast_mat_list <- function(reporting_triangle_list,
...
+      }
+    },
     reporting_triangle_list,
     delay_pmf_list,
-    MoreArgs = list(max_delay = max_delay, n = n),
     SIMPLIFY = FALSE
   )
 
 
+  non_null_indices <- which(!sapply(pt_nowcast_mat_list, is.null))
```
[the remark this answers is not in the document]
Comment (kaitejohnson) [reply R/generate_pt_nowcast_mat_list.R]: Refactored

File: R/estimate_dispersion.R [hunk 3]
```diff
@@ -74,9 +74,21 @@ estimate_dispersion <- function(
...
-  list_of_obs <- trunc_rep_tri_list[1:n]
+  # Truncate to only n nowcasts and extract only non-null elements of both lists
+  non_null_indices <- which(!sapply(pt_nowcast_mat_list[1:n], is.null))
+  n_iters <- length(non_null_indices)
+  list_of_ncs <- pt_nowcast_mat_list[non_null_indices]
+  list_of_obs <- trunc_rep_tri_list[non_null_indices]
+  if (n_iters == 0) {
+    cli_abort(
+      message = c(
+        "No retrospective nowcasts were generated, therefore ",
```
Comment (seabbs) [inline R/estimate_dispersion.R:85]:
```suggestion
        "No valid retrospective nowcasts were found, therefore ",
```

they are an input not generated here.

File: R/generate_pt_nowcast_mat_list.R [hunk 4]
```diff
@@ -64,16 +64,60 @@ generate_pt_nowcast_mat_list <- function(reporting_triangle_list,
...
     delay_pmf_list,
-    MoreArgs = list(max_delay = max_delay, n = n),
+    seq_along(reporting_triangle_list),
     SIMPLIFY = FALSE
   )
 
+  # After running, filter the results to find error indices
+  error_indices <- which(sapply(pt_nowcast_mat_list, is.null))
+  # Print summary
+  if (length(error_indices) == length(reporting_triangle_list)) {
```
Comment (seabbs) [inline R/generate_pt_nowcast_mat_list.R:103]: In terms of a user experience it would probably better if there was the summary of the errors followed by the detail vs the detail followed by a summary. I think it is fine to move forward like this but we may want a low priority issue to update this in the future.

File: R/generate_pt_nowcast_mat_list.R [hunk 5]
```diff
@@ -64,16 +64,60 @@ generate_pt_nowcast_mat_list <- function(reporting_triangle_list,
...
+    cli_abort(
+      message = c(sprintf(
+        "\nErrors occurred in all %s reporting triangles. Check if input triangles have valid data structure or contain zeros in the first column.", # nolint
+        length(reporting_triangle_list)
+      ))
+    )
+  } else if (length(error_indices) > 0) {
+    cat(sprintf("\nErrors occurred at indices: %s\n", toString(error_indices,
```
Comment (seabbs) [inline R/generate_pt_nowcast_mat_list.R:111]: What happened to using cli?

**Review by seabbs — DISMISSED** [review]
Functionality looks good. Is there a reason to move away from using cli for reporting?

File: R/generate_pt_nowcast_mat_list.R (hunk 5 above)
Comment (kaitejohnson) [reply R/generate_pt_nowcast_mat_list.R:111]: I have used cli for errors and warnings, but I thought the request here was just for a message (which, to my understanding is technically different from a warning). 

Could use `cli_alert()`?

File: R/generate_pt_nowcast_mat_list.R (hunk 4 above)
Comment (kaitejohnson) [reply R/generate_pt_nowcast_mat_list.R:103]: Yeah this prints within the iterator, in a similar way to stan where it tells you whats happening as its happening and then gives a summary. I prefer the errors being printed in order of the code running but not a strong preference. Will make an issue.

File: R/generate_pt_nowcast_mat_list.R [hunk 6]
```diff
@@ -64,16 +64,60 @@ generate_pt_nowcast_mat_list <- function(reporting_triangle_list,
...
-    generate_pt_nowcast_mat,
+    function(triangle, pmf, ind) {
+      result <- safe_generate_pt_nowcast_mat(
+        reporting_triangle = triangle,
+        delay_pmf = pmf,
+        n = n,
+        max_delay = max_delay
+      )
+      if (!is.null(result$error)) {
+        # Print the index and the error message
```
[the remark this answers is not in the document]
Comment (kaitejohnson) [reply R/generate_pt_nowcast_mat_list.R:81]: added a suffix

File: tests/testthat/test-generate_pt_nowcast_mat_list.R [hunk 7]
```diff
@@ -144,6 +144,70 @@ test_that("generate_pt_nowcast_mat_list identical-sized matrices work", {
...
+    byrow = TRUE
+  )
+
+  retro_rts_list <- list(test_triangle_1, test_triangle_2, triangle3)
+
+  result <- expect_message(generate_pt_nowcast_mat_list(retro_rts_list))
+  expect_null(result[[3]])
+})
+
+test_that("Function errors if only contains triangles with first column 0", { # nolint
```
Comment (seabbs) [inline tests/testthat/test-generate_pt_nowcast_mat_list.R:167]: function name here and elsewher.

**Comment by seabbs** [comment]
Flagging some outstanding issues here to do with consistency that I think needs another look ahead of a full review.

File: tests/testthat/test-generate_pt_nowcast_mat_list.R (hunk 7 above)
Comment (kaitejohnson) [reply tests/testthat/test-generate_pt_nowcast_mat_list.R]: I changed all the testthat strings to use the function name in the files edited here. I think we should streamline in a separate issue for remainder.

**Comment by kaitejohnson** [comment]
> Flagging some outstanding issues here to do with consistency that I think needs another look ahead of a full review.

Is this referring to the function names in the testing strings or something else?

**Review by seabbs — APPROVED** [review]
